File: BibliaAppExpo/src/services/daily-reading.service.ts
import { apiClient } from './api.client';

// ========== Types ==========

export interface DailyReading {
  id: string;
  date: string;
  title: string;
  badge: string;
  imageUrl: string | null;
  bookId: string;
  bookName: string;
  chapterNumber: number;
  verseNumbers: number[];
  readingText: string;
  biblicalReference: string;
  officialReflection: string | null;
}

export interface WeekReadings {
  readings: DailyReading[];
  startDate: string;
  endDate: string;
}

export interface ReadingHistory {
  id: string;
  readingId: string;
  date: string;
  completedAt: string;
}

// ========== API Service ==========

class DailyReadingService {
  /**
   * Get today's reading
   */
  async getTodayReading(): Promise<DailyReading> {
    const response = await apiClient.get<DailyReading>('/daily-reading/today');
    return response;
  }

  /**
   * Get reading for a specific date
   */
  async getReadingByDate(date: string): Promise<DailyReading> {
    const response = await apiClient.get<DailyReading>(`/daily-reading/${date}`);
    return response;
  }

  /**
   * Get readings for a week
   */
  async getWeekReadings(startDate?: string): Promise<WeekReadings> {
    const params = startDate ? `?startDate=${startDate}` : '';
    const response = await apiClient.get<WeekReadings>(`/daily-reading/week${params}`);
    return response;
  }

  /**
   * Mark a reading as completed
   */
  async markAsRead(readingId: string): Promise<ReadingHistory> {
    const response = await apiClient.post<ReadingHistory>(`/daily-reading/${readingId}/mark-read`, {});
    return response;
  }

  /**
   * Get user's reading history
   */
  async getReadingHistory(options?: {
    startDate?: string;
    endDate?: string;
    page?: number;
    size?: number;
  }): Promise<{ history: ReadingHistory[]; totalDays: number; streak: number }> {
    const params = new URLSearchParams();
    if (options?.startDate) params.append('startDate', options.startDate);
    if (options?.endDate) params.append('endDate', options.endDate);
    if (options?.page) params.append('page', options.page.toString());
    if (options?.size) params.append('size', options.size.toString());

    const queryString = params.toString();
    const response = await apiClient.get<{ history: ReadingHistory[]; totalDays: number; streak: number }>(
      `/daily-reading/history${queryString ? `?${queryString}` : ''}`
    );
    return response;
  }

  /**
   * Get current reading streak
   */
  async getStreak(): Promise<{ currentStreak: number; longestStreak: number; totalDaysRead: number }> {
    const response = await apiClient.get<{ currentStreak: number; longestStreak: number; totalDaysRead: number }>(
      '/daily-reading/streak'
    );
    return response;
  }
}

export const dailyReadingService = new DailyReadingService();

File: BibliaBackend/src/main/java/com/bibliacatolica/api/application/service/DailyReadingService.java
package com.bibliacatolica.api.application.service;

import com.bibliacatolica.api.domain.exception.ResourceNotFoundException;
import com.bibliacatolica.api.domain.model.DailyReading;
import com.bibliacatolica.api.domain.model.ReadingHistory;
import com.bibliacatolica.api.domain.port.in.DailyReadingUseCase;
import com.bibliacatolica.api.domain.port.out.DailyReadingRepositoryPort;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Servicio de aplicación que implementa los casos de uso de lectura diaria
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class DailyReadingService implements DailyReadingUseCase {

    private final DailyReadingRepositoryPort dailyReadingRepository;

    @Override
    @Transactional(readOnly = true)
    public DailyReading getTodayReading() {
        log.debug("Obteniendo lectura del día");
        return getReadingByDate(LocalDate.now());
    }

    @Override
    @Transactional(readOnly = true)
    public DailyReading getReadingByDate(LocalDate date) {
        log.debug("Obteniendo lectura para fecha: {}", date);

        // Intentar obtener lectura existente
        Optional<DailyReading> existingReading = dailyReadingRepository.findByDate(date);

        if (existingReading.isPresent()) {
            return existingReading.get();
        }

        // Si no existe, generar una lectura dinámica basada en la fecha
        // Usamos el día del año para seleccionar una lectura del pool
        return generateDynamicReading(date);
    }

    /**
     * Genera una lectura dinámica para fechas que no tienen lectura predefinida
     * Usa el día del año para seleccionar consistentemente la misma lectura para la misma fecha
     */
    private DailyReading generateDynamicReading(LocalDate date) {
        log.debug("Generando lectura dinámica para fecha: {}", date);

        // Obtener todas las lecturas disponibles
        List<DailyReading> allReadings = dailyReadingRepository.findAll();

        if (allReadings.isEmpty()) {
            throw new ResourceNotFoundException("Lectura diaria", date.toString());
        }

        // Usar el día del año para seleccionar consistentemente
        int dayOfYear = date.getDayOfYear();
        int index = dayOfYear % allReadings.size();

        DailyReading baseReading = allReadings.get(index);

        // Crear una copia con la fecha solicitada
        return DailyReading.builder()
                .id(UUID.nameUUIDFromBytes(("reading-" + date.toString()).getBytes()))
                .date(date)
                .title(baseReading.getTitle())
                .badge(baseReading.getBadge())
                .imageUrl(baseReading.getImageUrl())
                .bookId(baseReading.getBookId())
                .bookName(baseReading.getBookName())
                .chapterNumber(baseReading.getChapterNumber())
                .verseNumbers(baseReading.getVerseNumbers())
                .readingText(baseReading.getReadingText())
                .officialReflection(baseReading.getOfficialReflection())
                .build();
    }

    @Override
    @Transactional(readOnly = true)
    public List<DailyReading> getReadingsForWeek(LocalDate startDate) {
        log.debug("Obteniendo lecturas de la semana desde: {}", startDate);
        LocalDate endDate = startDate.plusDays(6);
        return dailyReadingRepository.findByDateBetween(startDate, endDate);
    }

    @Override
    @Transactional
    public ReadingHistory markAsRead(UUID userId, UUID readingId) {
        log.info("Usuario {} marcando lectura {} como leída", userId, readingId);

        // Verificar que la lectura existe
        dailyReadingRepository.findByDate(LocalDate.now())
                .filter(r -> r.getId().equals(readingId))
                .orElseThrow(() -> new ResourceNotFoundException("Lectura diaria", readingId.toString()));

        // Verificar si ya existe un registro
        Optional<ReadingHistory> existing = dailyReadingRepository
                .findHistoryByUserAndReading(userId, readingId);

        if (existing.isPresent()) {
            return existing.get();
        }

        // Crear nuevo registro de historial
        ReadingHistory history = ReadingHistory.builder()
                .id(UUID.randomUUID())
                .userId(userId)
                .dailyReadingId(readingId)
                .readAt(LocalDateTime.now())
                .build();

        return dailyReadingRepository.saveHistory(history);
    }

    @Override
    @Transactional
    public ReadingHistory saveReflection(UUID userId, UUID readingId, String reflection) {
        log.info("Usuario {} guardando reflexión para lectura {}", userId, readingId);

        // Buscar o crear historial
        ReadingHistory history = dailyReadingRepository
                .findHistoryByUserAndReading(userId, readingId)
                .orElseGet(() -> {
                    // Crear nuevo registro si no existe
                    return dailyReadingRepository.saveHistory(
                            ReadingHistory.builder()
                                    .id(UUID.randomUUID())
                                    .userId(userId)
                                    .dailyReadingId(readingId)
                                    .readAt(LocalDateTime.now())
                                    .build()
                    );
                });

        // Actualizar con la reflexión
        ReadingHistory updated = history.withReflection(reflection);
        return dailyReadingRepository.updateHistory(updated);
    }

    @Override
    @Transactional(readOnly = true)
    public List<ReadingHistoryDetail> getUserReadingHistory(UUID userId) {
        log.debug("Obteniendo historial de lecturas para usuario: {}", userId);

        List<ReadingHistory> histories = dailyReadingRepository.findHistoryByUserId(userId);

        return histories.stream()
                .map(history -> {
                    // Obtener la lectura asociada
                    // En una implementación real, esto debería optimizarse con un JOIN
                    DailyReading reading = null; // Se obtendría de la BD
                    return new ReadingHistoryDetail(
                            reading,
                            history,
                            history.hasReflection()
                    );
                })
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public List<ReadingHistoryDetail> getUserReadingHistoryByMonth(UUID userId, int year, int month) {
        log.debug("Obteniendo historial del mes {}/{} para usuario: {}", month, year, userId);

        LocalDate startDate = LocalDate.of(year, month, 1);
        LocalDate endDate = startDate.plusMonths(1).minusDays(1);

        List<ReadingHistory> histories = dailyReadingRepository
                .findHistoryByUserIdAndDateRange(userId, startDate, endDate);

        return histories.stream()
                .map(history -> new ReadingHistoryDetail(null, history, history.hasReflection()))
                .collect(Collectors.toList());
    }

    @Override
    @Transactional(readOnly = true)
    public boolean hasReadToday(UUID userId) {
        log.debug("Verificando si usuario {} ha leído hoy", userId);

        return dailyReadingRepository.findByDate(LocalDate.now())
                .map(reading -> dailyReadingRepository.hasUserReadToday(userId, reading.getId()))
                .orElse(false);
    }

    @Override
    @Transactional
    public DailyReading saveDailyReading(DailyReading dailyReading) {
        log.info("CMS: Guardando lectura litúrgica para fecha {}", dailyReading.getDate());
        // Si el id es nulo, asignamos uno aleatorio
        if (dailyReading.getId() == null) {
            dailyReading = baseReadingWithGeneratedId(dailyReading);
        }
        return dailyReadingRepository.save(dailyReading);
    }

    private DailyReading baseReadingWithGeneratedId(DailyReading reading) {
        return DailyReading.builder()
                .id(UUID.randomUUID())
                .date(reading.getDate())
                .title(reading.getTitle())
                .badge(reading.getBadge())
                .imageUrl(reading.getImageUrl())
                .bookId(reading.getBookId())
                .bookName(reading.getBookName())
                .chapterNumber(reading.getChapterNumber())
                .verseNumbers(reading.getVerseNumbers())
                .readingText(reading.getReadingText())
                .officialReflection(reading.getOfficialReflection())
                .build();
    }

    @Override
    @Transactional
    public void deleteDailyReadingByDate(LocalDate date) {
        log.info("CMS: Eliminando lectura litúrgica de fecha {}", date);
        dailyReadingRepository.findByDate(date).ifPresent(reading -> {
            dailyReadingRepository.delete(reading.getId());
        });
    }
}

